File: 747-min-cost-climbing-stairs/min-cost-climbing-stairs.js
/**
 * @param {number[]} cost
 * @return {number}
 */
var minCostClimbingStairs = function (cost) {

    // one or two 
    // f(3) = min(f(2)+cost[2] , f(1)+cost[1])
    // f(1) = f(0);
    // f(2) = f(0);
    let map = new Map();
    return DP(cost.length);

    function DP(floor) {

        if (floor == 0 || floor == 1) {
            return 0;
        }
        if(map.has(floor)){
            return map.get(floor);
        }
        map.set(floor,Math.min(DP(floor - 1) + cost[floor - 1], DP(floor - 2) + cost[floor - 2]));
        return map.get(floor);
    }

};

File: 747-min-cost-climbing-stairs/min-cost-climbing-stairs.java
class Solution {
    public int minCostClimbingStairs(int[] cost) {
        int stairs = cost.length;
        int[] mincost = new int[stairs];

        // climb 1 or 2 steps
        // start with index 0 or 1
        mincost[0] = cost[0];
        mincost[1] = cost[1];

        for (int i = 2; i < stairs; i++) {
            mincost[i] = cost[i] + Math.min(mincost[i - 1], mincost[i - 2]);
        }

        return Math.min(mincost[stairs - 1], mincost[stairs - 2]);
    }
}